pkg/progressreader/progressstatus.go

```diff
@@ -0,0 +1,71 @@
+package progressreader
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sync"
+)
+
+type ProgressStatus struct {
+	sync.Mutex
+	c         chan struct{}
+	observers []io.Writer
+	history   bytes.Buffer
+}
+
+func NewProgressStatus() *ProgressStatus {
+	return &ProgressStatus{
+		c:         make(chan struct{}),
+		observers: []io.Writer{},
+	}
+}
+
+func (ps *ProgressStatus) Write(p []byte) (n int, err error) {
+	ps.Lock()
+	defer ps.Unlock()
+	ps.history.Write(p)
+	for _, w := range ps.observers {
+		// copy paste from MultiWriter, replaced return with continue
+		n, err = w.Write(p)
+		if err != nil {
+			continue
+		}
+		if n != len(p) {
+			err = io.ErrShortWrite
+			continue
+		}
+	}
+	return len(p), nil
+}
+
+func (ps *ProgressStatus) AddObserver(w io.Writer) {
+	ps.Lock()
+	defer ps.Unlock()
+	w.Write(ps.history.Bytes())
+	ps.observers = append(ps.observers, w)
+}
+
+func (ps *ProgressStatus) Done() {
+	ps.Lock()
+	close(ps.c)
+	ps.history.Reset()
+	ps.Unlock()
+}
+
+func (ps *ProgressStatus) Wait(w io.Writer, msg []byte) error {
+	ps.Lock()
+	channel := ps.c
+	ps.Unlock()
+
+	if channel == nil {
+		// super defensive, not user facing message, channel was not allocated
```

Comment:
If this isn't user facing, and this error appears to be swallowed, why throw an error at all?

Is the expectation that channel would be nil after Done is called? I don't see `pc.c` getting set to nil.